PurchasesCoreSwift/Public/EntitlementInfo.swift

```diff
@@ -0,0 +1,346 @@
+        let unsubscribeDetectedAtString = productData["unsubscribe_detected_at"] as? String
+        let billingIssuesDetectedAtString = productData["billing_issues_detected_at"] as? String
+        let ownershipType = productData["ownership_type"] as? String
+
+        let store = Self.parseStore(store: storeString)
+        let expirationDate = Self.parseDate(dateString: productExpiresDateString, withDateFormatter: dateFormatter)
+        let unsubscribeDetectedAt = Self.parseDate(dateString: unsubscribeDetectedAtString,
+                                                   withDateFormatter: dateFormatter)
+        let billingIssueDetectedAt = Self.parseDate(dateString: billingIssuesDetectedAtString,
+                                                    withDateFormatter: dateFormatter)
+        let entitlementExpiresDate = Self.parseDate(dateString: entitlementExpiresDateString,
+                                                    withDateFormatter: dateFormatter)
+
+        self.store = store
+        self.expirationDate = expirationDate
+        self.unsubscribeDetectedAt = unsubscribeDetectedAt
+        self.billingIssueDetectedAt = billingIssueDetectedAt
+        self.identifier = entitlementId
+        self.productIdentifier = productIdString!
+
+        let isSandbox: Bool
+        if maybeSandbox?.responds(to: #selector(getter: NSNumber.boolValue)) ?? false,
+           let unwrapped = maybeSandbox as? NSNumber {
+            isSandbox = unwrapped.boolValue
+        } else if maybeSandbox?.responds(to: #selector(getter: NSString.boolValue)) ?? false,
+                  let unwrapped = maybeSandbox as? NSString {
+            isSandbox = unwrapped.boolValue
+        } else {
+            isSandbox = false
+        }
+
+        self.isSandbox = isSandbox
+        self.isActive = Self.isDateActive(expirationDate: entitlementExpiresDate, forRequestDate: requestDate)
+        self.periodType = Self.parsePeriodType(periodType: periodTypeString)
+        self.latestPurchaseDate = Self.parseDate(dateString: entitlementPurchaseDateString,
+                                                 withDateFormatter: dateFormatter)
+        self.originalPurchaseDate = Self.parseDate(dateString: originalPurchaseDateString,
+                                                   withDateFormatter: dateFormatter)
+        self.ownershipType = Self.parseOwnershipType(ownershipType: ownershipType)
+        self.willRenew = Self.willRenewWithExpirationDate(expirationDate: expirationDate,
+                                                          store: store,
+                                                          unsubscribeDetectedAt: unsubscribeDetectedAt,
+                                                          billingIssueDetectedAt: billingIssueDetectedAt)
+    }
+
+    class func parseDate(dateString: String?, withDateFormatter dateFormatter: DateFormatter) -> Date? {
```

Comment:
I'd love to move all of these into a builder one day (obviously after the migration)

Reply:
we should make them private, though, since they're not meant to be used outside the scope of this file

Reply:
✅